File: mobile/customer-app/app/src/main/java/com/sukashawarma/customer/data/CartStore.kt
package com.sukashawarma.customer.data

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/** Batas jumlah per baris. Cerminan `JUMLAH_MAKS_PER_ITEM` di gateway. */
const val JUMLAH_MAKS_PER_ITEM = 99

/** Batas panjang catatan. Gateway memotong di angka yang sama. */
const val PANJANG_MAKS_CATATAN = 200

@Serializable
data class CartLine(
    val menuItemId: String,
    val nama: String,
    /**
     * Rupiah penuh. Rupiah tidak punya satuan pecahan, dan gateway
     * membandingkan `unit_price` dengan katalog memakai kesamaan PERSIS.
     * Menyimpannya sebagai Double akan membuka celah galat pembulatan tepat
     * di titik yang paling tidak boleh meleset.
     */
    val hargaSatuan: Long,
    val jumlah: Int,
    val catatan: String? = null
)

@Serializable
private data class IsiKeranjang(
    val outletId: String? = null,
    val baris: List<CartLine> = emptyList()
)

/** Cara keranjang bertahan lintas proses. Dipisah supaya bisa diuji tanpa Android. */
interface CartPersistence {
    fun muat(): String?
    fun simpan(isi: String)
}

/**
 * Keranjang pelanggan.
 *
 * **Kenapa bukan DataStore** (rencana menyebut DataStore): antarmuka yang
 * rencana itu sendiri tentukan bersifat sinkron (`isi(): List<CartLine>`,
 * `subtotal(): Long`), sedangkan DataStore seluruhnya asinkron. Membungkus
 * DataStore di balik pemanggilan sinkron berarti memblokir thread utama atau
 * menyimpan salinan bayangan di memori -- dua-duanya lebih buruk daripada
 * SharedPreferences, yang memang dirancang untuk data sekecil ini dan sudah
 * sinkron sejak awal. Keranjang tetap bertahan lintas proses, sesuai §5.4.
 */
class CartStore internal constructor(private val penyimpan: CartPersistence?) {

    private val json = Json { ignoreUnknownKeys = true }
    private var isi: IsiKeranjang = muatAwal()

    private fun muatAwal(): IsiKeranjang {
        val mentah = penyimpan?.muat() ?: return IsiKeranjang()
        // Keranjang rusak tidak boleh mematikan aplikasi saat dibuka. Lebih
        // baik keranjang kosong daripada layar yang tidak pernah muncul.
        return runCatching { json.decodeFromString<IsiKeranjang>(mentah) }
            .getOrElse { IsiKeranjang() }
    }

    private fun tulis() {
        penyimpan?.simpan(json.encodeToString(isi))
    }

    fun outletId(): String? = isi.outletId

    /**
     * Menetapkan outlet keranjang.
     *
     * Berpindah outlet MENGOSONGKAN keranjang, dan mengembalikan `true` bila
     * itu terjadi supaya layar bisa memberi tahu. `menu_item_id` bersifat
     * per-outlet (katalog disaring `outlet_id`), jadi membawa isi keranjang
     * outlet A ke outlet B menghasilkan pesanan yang PASTI ditolak gateway
     * dengan "tidak_ada" -- tepat di titik pembayaran, bukan di sini.
     */
    fun pakaiOutlet(outletId: String): Boolean {
        if (isi.outletId == outletId) return false
        val adaIsi = isi.baris.isNotEmpty()
        isi = IsiKeranjang(outletId = outletId, baris = emptyList())
        tulis()
        return adaIsi
    }

    /**
     * Menambahkan item.
     *
     * Baris digabung hanya bila id DAN catatannya sama. Catatan berbeda =
     * instruksi dapur berbeda, jadi harus jadi baris sendiri; menggabungkannya
     * akan membuat salah satu catatan hilang diam-diam.
     */
    fun tambah(
        menuItemId: String,
        nama: String,
        hargaSatuan: Long,
        jumlah: Int,
        catatan: String?
    ) {
        val catatanBersih = rapikanCatatan(catatan)
        val tambahan = jumlah.coerceIn(1, JUMLAH_MAKS_PER_ITEM)

        val baris = isi.baris.toMutableList()
        val posisi = baris.indexOfFirst {
            it.menuItemId == menuItemId && it.catatan == catatanBersih
        }

        if (posisi >= 0) {
            val lama = baris[posisi]
            baris[posisi] = lama.copy(
                jumlah = (lama.jumlah + tambahan).coerceAtMost(JUMLAH_MAKS_PER_ITEM)
            )
        } else {
            baris.add(
                CartLine(
                    menuItemId = menuItemId,
                    nama = nama,
                    hargaSatuan = hargaSatuan,
                    jumlah = tambahan,
                    catatan = catatanBersih
                )
            )
        }

        isi = isi.copy(baris = baris)
        tulis()
    }

    /** Menaikkan atau menurunkan jumlah. Turun sampai nol menghapus barisnya. */
    fun ubahJumlah(index: Int, delta: Int) {
        val baris = isi.baris.toMutableList()
        if (index !in baris.indices) return

        val baru = baris[index].jumlah + delta
        if (baru <= 0) {
            baris.removeAt(index)
        } else {
            baris[index] = baris[index].copy(
                jumlah = baru.coerceAtMost(JUMLAH_MAKS_PER_ITEM)
            )
        }

        isi = isi.copy(baris = baris)
        tulis()
    }

    fun hapus(index: Int) {
        val baris = isi.baris.toMutableList()
        if (index !in baris.indices) return
        baris.removeAt(index)
        isi = isi.copy(baris = baris)
        tulis()
    }

    /**
     * Membuang semua baris untuk satu menu.
     *
     * Dipakai saat gateway melaporkan item habis atau sudah tidak ada. Satu
     * menu bisa menempati beberapa baris (catatan berbeda), jadi menghapus
     * berdasarkan indeks akan menyisakan sebagian -- dan checkout gagal lagi
     * dengan keluhan yang sama persis.
     */
    fun hapusMenuItem(menuItemId: String) {
        isi = isi.copy(baris = isi.baris.filterNot { it.menuItemId == menuItemId })
        tulis()
    }

    /**
     * Menyetel harga satu menu ke harga terbaru dari gateway.
     *
     * Menerima harga baru adalah keputusan pelanggan, bukan sesuatu yang boleh
     * terjadi diam-diam: aplikasi memanggil ini hanya setelah harga barunya
     * ditampilkan dan pelanggan menekan tombolnya sendiri.
     */
    fun perbaruiHarga(menuItemId: String, hargaBaru: Long) {
        isi = isi.copy(
            baris = isi.baris.map {
                if (it.menuItemId == menuItemId) it.copy(hargaSatuan = hargaBaru) else it
            }
        )
        tulis()
    }

    fun isi(): List<CartLine> = isi.baris

    fun kosongkan() {
        isi = isi.copy(baris = emptyList())
        tulis()
    }

    fun subtotal(): Long = isi.baris.sumOf { it.hargaSatuan * it.jumlah }

    fun jumlahPorsi(): Int = isi.baris.sumOf { it.jumlah }

    companion object {
        /** Keranjang tanpa penyimpanan. Untuk test. */
        fun diMemori(): CartStore = CartStore(null)

        fun persisten(context: Context): CartStore =
            CartStore(PrefsCartPersistence(context.applicationContext))
    }
}

/**
 * Memotong catatan di 200 karakter DI APLIKASI, bukan hanya di gateway.
 *
 * Gateway juga memotongnya, tapi memotong lebih awal berarti pelanggan
 * melihat batasnya sendiri -- bukan mengetik seratus karakter lagi lalu
 * kehilangannya tanpa pernah diberi tahu.
 *
 * `|NOTE|` dibuang karena konvensi struk dapur memakai penanda itu untuk
 * memisahkan nama item dari catatan; catatan yang memuatnya akan merusak
 * cetakan dapur. Gateway juga membersihkannya, ini lapisan kedua.
 */
fun rapikanCatatan(catatan: String?): String? {
    val bersih = catatan?.replace("|NOTE|", " ")?.trim()?.take(PANJANG_MAKS_CATATAN)
    return if (bersih.isNullOrBlank()) null else bersih
}

private class PrefsCartPersistence(context: Context) : CartPersistence {
    private val prefs = context.getSharedPreferences("suka_customer_cart", Context.MODE_PRIVATE)
    override fun muat(): String? = prefs.getString("isi", null)
    override fun simpan(isi: String) {
        prefs.edit().putString("isi", isi).apply()
    }
}

File: mobile/customer-ios-app/SukaShawarma/Data/CartStore.swift
import Foundation
import Observation

/// Batas jumlah per baris. Cerminan `JUMLAH_MAKS_PER_ITEM` di gateway.
let jumlahMaksPerItem = 99

/// Batas panjang catatan. Gateway memotong di angka yang sama.
let panjangMaksCatatan = 200

struct CartTopping: Codable, Sendable, Equatable, Hashable {
    let menuItemId: String
    let nama: String
    var hargaSatuan: Int64
}

struct CartLine: Codable, Sendable, Equatable {
    let menuItemId: String
    let nama: String
    /// Rupiah penuh (Int64), BUKAN Double: gateway membandingkan `unit_price`
    /// dengan katalog memakai kesamaan PERSIS.
    var hargaSatuan: Int64
    var jumlah: Int
    var catatan: String?
    var toppings: [CartTopping] = []
}

private struct IsiKeranjang: Codable {
    var outletId: String?
    var baris: [CartLine] = []
}

/// Cara keranjang bertahan lintas peluncuran. Dipisah agar bisa diuji.
protocol CartPersistence: AnyObject {
    func muat() -> String?
    func simpan(_ isi: String)
}

/// Padanan `data/CartStore.kt`. `@Observable` supaya layar (Fase 2+) ikut
/// diperbarui saat isi keranjang berubah. Aturan bisnisnya identik dengan
/// Android dan dijaga oleh `CartStoreTests` (porting `CartStoreTest.kt`).
@MainActor
@Observable
final class CartStore {
    @ObservationIgnored private let penyimpan: CartPersistence?
    private var keadaan: IsiKeranjang

    init(penyimpan: CartPersistence?) {
        self.penyimpan = penyimpan
        // Keranjang rusak tidak boleh mematikan aplikasi saat dibuka.
        if let mentah = penyimpan?.muat(),
           let data = mentah.data(using: .utf8),
           let terurai = try? JSONDecoder().decode(IsiKeranjang.self, from: data) {
            keadaan = terurai
        } else {
            keadaan = IsiKeranjang()
        }
    }

    /// Keranjang tanpa penyimpanan. Untuk uji.
    static func diMemori() -> CartStore { CartStore(penyimpan: nil) }

    static func persisten() -> CartStore { CartStore(penyimpan: DefaultsCartPersistence()) }

    private func tulis() {
        guard let penyimpan, let data = try? JSONEncoder().encode(keadaan),
              let teks = String(data: data, encoding: .utf8) else { return }
        penyimpan.simpan(teks)
    }

    func outletId() -> String? { keadaan.outletId }

    /// Berpindah outlet MENGOSONGKAN keranjang dan mengembalikan `true` bila
    /// itu terjadi. `menu_item_id` bersifat per-outlet — membawa isi keranjang
    /// outlet A ke outlet B pasti ditolak gateway tepat di titik bayar.
    @discardableResult
    func pakaiOutlet(_ outletId: String) -> Bool {
        if keadaan.outletId == outletId { return false }
        let adaIsi = !keadaan.baris.isEmpty
        keadaan = IsiKeranjang(outletId: outletId, baris: [])
        tulis()
        return adaIsi
    }

    /// Baris digabung hanya bila id, catatan, DAN topping-nya sama persis.
    func tambah(menuItemId: String, nama: String, hargaSatuan: Int64, jumlah: Int,
                catatan: String?, toppings: [CartTopping] = []) {
        let catatanBersih = rapikanCatatan(catatan)
        let tambahan = min(max(jumlah, 1), jumlahMaksPerItem)

        if let posisi = keadaan.baris.firstIndex(where: {
            $0.menuItemId == menuItemId && $0.catatan == catatanBersih && $0.toppings == toppings
        }) {
            keadaan.baris[posisi].jumlah = min(keadaan.baris[posisi].jumlah + tambahan, jumlahMaksPerItem)
        } else {
            keadaan.baris.append(CartLine(menuItemId: menuItemId, nama: nama, hargaSatuan: hargaSatuan,
                                          jumlah: tambahan, catatan: catatanBersih, toppings: toppings))
        }
        tulis()
    }

    /// Menaikkan atau menurunkan jumlah. Turun sampai nol menghapus barisnya.
    func ubahJumlah(index: Int, delta: Int) {
        guard keadaan.baris.indices.contains(index) else { return }
        let baru = keadaan.baris[index].jumlah + delta
        if baru <= 0 {
            keadaan.baris.remove(at: index)
        } else {
            keadaan.baris[index].jumlah = min(baru, jumlahMaksPerItem)
        }
        tulis()
    }

    /// Menghapus sub-item topping tertentu dari suatu baris pesanan.
    func hapusTopping(index: Int, toppingMenuItemId: String) {
        guard keadaan.baris.indices.contains(index) else { return }
        keadaan.baris[index].toppings.removeAll { $0.menuItemId == toppingMenuItemId }
        tulis()
    }

    func hapus(index: Int) {
        guard keadaan.baris.indices.contains(index) else { return }
        keadaan.baris.remove(at: index)
        tulis()
    }

    /// Membuang SEMUA baris satu menu (sebagai item utama maupun topping).
    /// Dipakai saat gateway melaporkan item habis/tidak ada.
    func hapusMenuItem(_ menuItemId: String) {
        keadaan.baris = keadaan.baris
            .filter { $0.menuItemId != menuItemId }
            .map { line in
                var l = line
                l.toppings.removeAll { $0.menuItemId == menuItemId }
                return l
            }
        tulis()
    }

    /// Menyetel harga satu menu ke harga terbaru dari gateway. Dipanggil HANYA
    /// setelah pelanggan melihat harga baru dan menyetujuinya sendiri.
    func perbaruiHarga(menuItemId: String, hargaBaru: Int64) {
        keadaan.baris = keadaan.baris.map { line in
            var l = line
            if l.menuItemId == menuItemId { l.hargaSatuan = hargaBaru }
            l.toppings = l.toppings.map { t in
                var top = t
                if top.menuItemId == menuItemId { top.hargaSatuan = hargaBaru }
                return top
            }
            return l
        }
        tulis()
    }

    func isi() -> [CartLine] { keadaan.baris }

    func kosongkan() {
        keadaan.baris = []
        tulis()
    }

    func subtotal() -> Int64 {
        keadaan.baris.reduce(0) { total, line in
            let perPorsi = line.hargaSatuan + line.toppings.reduce(0) { $0 + $1.hargaSatuan }
            return total + perPorsi * Int64(line.jumlah)
        }
    }

    func jumlahPorsi() -> Int { keadaan.baris.reduce(0) { $0 + $1.jumlah } }
}

/// Memotong catatan di 200 karakter DI APLIKASI (pelanggan melihat batasnya
/// sendiri) dan membuang `|NOTE|` yang merusak struk dapur.
func rapikanCatatan(_ catatan: String?) -> String? {
    guard let catatan else { return nil }
    let bersih = String(catatan.replacingOccurrences(of: "|NOTE|", with: " ")
        .trimmingCharacters(in: .whitespacesAndNewlines)
        .prefix(panjangMaksCatatan))
    return bersih.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : bersih
}

private final class DefaultsCartPersistence: CartPersistence {
    private let defaults = UserDefaults(suiteName: "suka_customer_cart") ?? .standard
    func muat() -> String? { defaults.string(forKey: "isi") }
    func simpan(_ isi: String) { defaults.set(isi, forKey: "isi") }
}
